exporter/exporterhelper/internal/persistent_queue.go

```diff
@@ -205,20 +205,25 @@ func (pq *persistentQueue[T]) putInternal(ctx context.Context, req T) error {
 	}
 
 	itemKey := getItemKey(pq.writeIndex)
-	pq.writeIndex++
+	newIndex := pq.writeIndex + 1
 
 	reqBuf, err := pq.marshaler(req)
 	if err != nil {
 		return err
 	}
-	err = pq.client.Batch(ctx,
-		storage.SetOperation(writeIndexKey, itemIndexToBytes(pq.writeIndex)),
-		storage.SetOperation(itemKey, reqBuf))
 
+	// Carry out a transaction where we both add the item and update the write index
+	setWriteIndexOp := storage.SetOperation(writeIndexKey, itemIndexToBytes(newIndex))
+	setItemOp := storage.SetOperation(itemKey, reqBuf)
+	if err := pq.client.Batch(ctx, setWriteIndexOp, setItemOp); err != nil {
+		return err
+	}
```

Comment:
I understand the return error, but why changing the way operations were constructed?

Reply:
I wanted to write this by assigning to `err` in the condition, which is more readable with a shorter statement imo. So I defined the operations separately. Do you think it's worse now?

Reply:
I think it just complicates the review, especially understanding what changed. Writing the code as you suggested makes sense in general, but here I feel that it complicates a bit (especially that you still shadow err). I would write maybe like:

```golang
ops := []storage.Operations{ storage.SetOperation(writeIndexKey, itemIndexToBytes(newIndex),  storage.SetOperation(itemKey, reqBuf)}
if err := pq.client.Batch(ctx, ops...); err != nil {
	return err
}
```

Reply:
Yeah, that looks better. Applied this change and also renamed the error variable to avoid shadowing.